### PR TITLE
Cav 367

### PR DESCRIPTION
**Pull request by mar36tin**

### CODE REVIEW

File: app/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapper.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    .onHalfOpen(logger.warn(s"Circuit breaker for ${configCB.serviceName} is half open")) //Scenario 2
+    .onClose {logger.warn(s"Circuit breaker for ${configCB.serviceName} has closed")} //Scenario 4
+    .onCallFailure(_ => logger.error(s"Circuit breaker for ${configCB.serviceName} recorded failed call"))
+    .onCallBreakerOpen {logger.error(s"Circuit breaker for ${configCB.serviceName} rejected call due to previous failures")}
+    .onCallTimeout {
+      elapsed =>
+        val duration = Duration.fromNanos(elapsed)
+        logger.error(s"Circuit breaker for ${configCB.serviceName} recorded failed call due to timeout after ${duration.toMillis}ms")
+    }
+}
```
Comment (johno1985) [inline app/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapper.scala:54]: no new line

File: test/uk/gov/hmrc/cipphonenumber/utils/TestActorSystem.scala [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+import org.scalatest.Suite
+
+object TestActorSystem {
+  val system: ActorSystem = ActorSystem("test")
+}
+
+trait TestActorSystem { self: Suite =>
+  implicit val system: ActorSystem        = TestActorSystem.system
+  implicit val materializer: Materializer = Materializer(TestActorSystem.system)
+}
```
Comment (johno1985) [inline test/uk/gov/hmrc/cipphonenumber/utils/TestActorSystem.scala:30]: no new line

**Review by johno1985 — COMMENTED** [review]
minor pedantic comments

**Review by johno1985 — APPROVED** [review]

**Review by johno1985 — DISMISSED** [review]

File: test/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapperSpec.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+          .withBody(Json.obj("phoneNumber" -> "test"))
+          .transform(_.withRequestFilter(AhcCurlRequestLogger()))
+          .execute[HttpResponse]
+      )
+
+      await(result).status shouldBe OK
+      circuitBreakers.circuitBreaker.isOpen shouldBe false
+      circuitBreakers.circuitBreaker.isClosed shouldBe true
+      circuitBreakers.circuitBreaker.isHalfOpen shouldBe false
+    }
```
Comment (brdgeth) [inline test/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapperSpec.scala:65]: is there a reason why the other status changes cannot be tested?  for example, can you mock a failed response, or a fault or an equivalent, to test what happens?

**Review by brdgeth — COMMENTED** [review]
one question about the unit testing

**Review by brdgeth — APPROVED** [review]

**Review by johno1985 — APPROVED** [review]

**Review by brdgeth — DISMISSED** [review]

File: app/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapper.scala [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    callTimeout = configCB.callTimeout, // time after which to consider a call a failure
+    resetTimeout = configCB.resetTimeout, // time after which to attempt to close the circuit
+    maxResetTimeout = configCB.maxResetTimeout, // max time after which to attempt to close the circuit
+    exponentialBackoffFactor = configCB.exponentialBackoffFactor, // exponential time gap e.g. 1 1.2 1.2+1.2*0.2
+    randomFactor = configCB.randomFactor //after calculation of the exponential back-off an additional random delay based on this factor is added, e.g. 0.2 adds up to 20% delay.
+  )(materializer.executionContext)
+    .onOpen(logger.warn(s"Circuit breaker for ${configCB.serviceName} opened and will not close for ${configCB.resetTimeout}"))
+    .onHalfOpen(logger.warn(s"Circuit breaker for ${configCB.serviceName} is half open"))
+    .onClose {logger.info(s"Circuit breaker for ${configCB.serviceName} has closed")}
+    .onCallFailure(_ => logger.error(s"Circuit breaker for ${configCB.serviceName} recorded failed call"))
```
Comment (johno1985) [inline app/uk/gov/hmrc/cipphonenumber/connectors/CircuitBreakerWrapper.scala:47]: I think these log errors should be log warnings also

**Review by johno1985 — COMMENTED** [review]
minor comment
rebase needed

**Review by johno1985 — APPROVED** [review]

**Review by brdgeth — APPROVED** [review]

**Pull request merged**